Inline osr

o- opened the pull request:
Not so well tested yet....

...but still worth a look.

o- committed:
extend the syntax of osr instruction to create additional frames

this allows osr to create additional frames. The syntax is:

osr l [...] (...) [...], (f,v,l)[var x = $, var y = 2], (...)[...]

There are several parts to it.

* The label l
* A list of conditions, eg. [x==2, x==3]
* The osr target, eg. (f,v,l)
* The varmap definition of the top frame, eg. [var x = 2]
* Finally (new) a comma separated list of extra frames.

An extra frame has the syntax `(f,v,l) [var x = $, ...]`
* The `()` part specifies return target of the additional frame
* The `[]` specifies the extra environment to create.
  It is similar to the normal varmap but needs to begin with an
  additional `var x = $` to specify which variable will hold the
  function result.

For example
   var x = 1
   call y = foo()
  after:

can be simulated by an osr extra frame:
  (f,v,after) [var y = $, var x = 1]

Each of those extra frame entries defines one continuation. Due to the
way we create continuations we have to in include the result variable
in the varmap with the special `$` syntax.

o- committed:
change transform_inline to be aware of checkpoints

To properly support osr across inlining the inlining pass
needs to know if there is a checkpoint after the function
call at the caller site.

This patch refactors compute_inlining_order to not just
return the order, but in fact an annotated graph that
contains all the information for the inliner to go forward.

At the core there is the `inlining_candidate` which contains
all information to apply one inlining. The inlinings are
performed recursively, all in one go.

At the end a new version is returned which can be added to
the first caller function.

Currently inlinings are skipped when changes to the osr
points would be needed.

o- committed:
extend transform_inline to handle osr

when creating inline_candidates we keep track of the safepoints
around the function call. We use the existing osr map to create
an extended version.

For example given:

  function caller
  version active
    var x = 1
    call y = bla (...)
    osr after_call ...  [var x = x]

We create the following frame map:

    (caller, active, after_call) [var y = $, var x = x]

This can be used to fixup the osrs in the inlinee. For example

  function bla
  version active
    osr cp_1 [] (bla, base, cp_1) [...]

Can be inlined as

    var x = 1
    osr cp_1 [] (bla, base, cp_1) [...], (caller, active, after_call) [var y = $, var x = x]
    osr after_call ...  [var x = x]

aviralg commented on transform_inline.ml:
```diff
@@ -103,8 +86,7 @@ let inline ({main; functions} as orig_prog : program) : program option =
-    let used_labels = extract_labels caller.instrs in
+  let replace_labels used_labels ({formals; instrs} as callee) =
     let callee_labels = LabelSet.elements (extract_labels callee.instrs) in
     let replacements = Edit.fresh_many used_labels callee_labels in
```
I think you can remove the `Edit` as it is already opened.

o- commented:
@aviralg to reiterate what gabriel said by mail: thanks a lot for the great work on the inliner. I hope you don't mind that I had to replace your inline order logic. The problem was that it was really hard to keep track of extra information while discovering the callgraph. That's why I chose to replace the inline order list by a more graph like structure that I can just traverse. The actual inline logic is still there and proved to be super robust.

aviralg reviewed (commented):
Its great! I found a couple of formatting nitpicks but I would leave them in the able hands of @mhyee .

o- commented:
There might be some rough edges, but it should be ready as a basis to talk about it...

The main idea is that our osr gets extended to the right by an optional list of additional continuations to create. for example `(a,b,c,x)[var y = 1]` creates the continuation `(a,b,c), x, (y->1)`. 

The most interesting part is [here](https://github.com/reactorlabs/sourir/pull/160/commits/54ff835d023d5c0341dd30e95fa6f8a8dd34af2e#diff-ddb3aa6ec7a2afe48b9b7aee59c17b26R181) where the current activation frame osr mapping is converted into an continuation creating entry that will be appended to the inlinee osr's.

aviralg commented:
@o- don't worry, I don't mind. Code has to be updated as the requirement changes.

mhyee commented on instr.ml:
```diff
@@ -413,7 +420,10 @@ class map = object (m)
+
+  method osr_frame ({var_map} as frame) = { frame with var_map = m#osr_var_map var_map; }
+  method osr_var_map = List.map m#osr_var_map_entry
+  method osr_var_map_entry = function
```
For consistency, I think this should be called `osr_def`.

o- replied:
maybe everything should be called map entry?

mhyee commented on instr.ml:
```diff
@@ -398,12 +404,13 @@ class map = object (m)
-    | Osr {label; cond; target; map} ->
+    | Osr {label; cond; target; var_map; frame_maps} ->
       Osr {
         label;
```
Sorry, but why do we *never* need a method for `label`? Likewise for `target`.

o- replied:
That's not a question I would deem valid. It's easy to add. My guideline is: Why do we need it *now*. My software engineering practice and also training has thought me to never ever add features just because they might come handy in the future. Like in this case I had to rewrite the existing code that was too flexible for it's own good. If I spent more than 2' trying to figure out how to abstract it even though I am sure we won't use it in the next 3 weeks that would be a waste. But what's worse is I add a badly tested (since unused) feature to the codebase. The more (untested) things in the codebase the harder it is to move it. I know that is now a completely overblown answer to this small point. But I thought it's a good situation to explain myself and why I sometimes remove things that I see as cruft.

gasche replied:
I am favorable to anti-over-engineering moves in general, but to be fair this is a case where the visitor has a *very* regular structure that always visits all arguments (I realize now that `op` is not visited, but that is a change you introduced), and I think moving away from this regular structure is wrong (before it was very simple to know what to do, and now people have to think about what to do in each case).

Besides, in that case visiting the label is precisely required if you want to freshen all the variables and labels of a piece of code. As for the `variable` case, we should have a generic `label` method that all three of `branch_label`, `goto_label` and `osr_label` defer to. I can send a PR after this one is merged.

o- replied:
fair enough, though I must note that the original #op method did not have access to the arguments and I am really wondering how you would change the operation independently from the args :)

the not having to think and seeing the pieces in isolation was imho the reason for the bug where we converted `x[i]` to `nil[i]`.

I am not arguing against a visitor here. The only thing I am saying is that it is far from clear to me what the granularity of the visits should be and that if I am not sure how to proceed with a feature we don't need currently I leave it to be implemented when a usecase arises.

gasche replied:
The issue was the mismatch between the type structure and what the parser allowed, not the visitor.

o- replied:
ok, sorry I think I escalated this into a bigger argument than necessary. apologies. I'll try to restore the visitor as it was.

mhyee commented on instr.ml:
```diff
@@ -413,7 +420,10 @@ class map = object (m)
       List (List.map m#expression es)
-  method osr_map = function
+
+  method osr_frame ({var_map} as frame) = { frame with var_map = m#osr_var_map var_map; }
```
Same question, why do we never need a method for `continuation_pos` or `continuation_res`?

o- replied:
see above. Even if I wanted I don't see a usecase for rewriting them.

mhyee commented on tests.ml:
```diff
@@ -1886,7 +1886,9 @@ let suite =
-   "parser5b">:: test_parse_disasm ("osr l [(x == y)] (f, v, l) [var x = x, var v, var x = (1+2)]\nl:\n");
+   "parser5b">:: test_parse_disasm ("osr l [(x == y)] (f, v, l) [var x = x, var v, var x = (1+2)]\n");
+   "parser5c">:: test_parse_disasm ("osr l [(x == y)] (f, v, l) [var x = x, var v, var x = (1+2)] (f,v,l,x) []\n");
+   "parser5d">:: test_parse_disasm ("osr l [(x == y)] (f, v, l) [var x = x, var v, var x = (1+2)] (f,v,l,x) [], (f,v,l,x) [var c = 4]\n");
```
I'll just put this comment here, but there's a mistake in the commit message. I was trying to use it to figure out the syntax of the new `osr` instruction. It was missing the list of conditions:

`osr l [...] (...) [...]  (f,v,l,x)[var y = 2], (...)[...]`

o- replied:
ups I'll try to fix it

mhyee replied:
The updated commit message is even better than before. Thanks!

mhyee commented on eval.ml:
```diff
@@ -372,25 +372,40 @@ let reduce conf =
     else begin
-      let osr_env, heap' = build_osr_frame map conf.env conf.heap in
+      let osr_env, heap = build_osr_frame var_map conf.env conf.heap in
+      let create_extra_frames (heap, frames) {continuation_pos; continuation_res; var_map} =
```
For consistency, I suppose this should be called `build_extra_frames`.

Also, can you move the definition up, to where `build_call_frame` and `build_osr_frame` are defined? That will keep this section a little cleaner.

mhyee replied:
I guess the `List.fold_left ...` part should also be moved out, so we just have a single line

```ocaml
let heap, extra_frames = build_extra_frames frame_maps in
```

(Plus whatever extra args it needs.)

o- replied:
sounds good

mhyee commented on eval.ml:
```diff
@@ -372,25 +372,40 @@ let reduce conf =
         pc = resolve_osr instrs label;
         env = osr_env;
-        heap = heap';
+        heap;
```
I'd prefer this to be more explicit (like `instrs` below): `heap = heap;`

o- replied:
that's kind of a very stylistic request, not?

mhyee replied:
Sure, it's not a big deal either way.

mhyee commented on transform_assumption.ml:
```diff
@@ -210,23 +210,23 @@ let hoist_assumption : transform_instructions = fun ({instrs; _} as inp) ->
-          | Osr {label; cond; target; map} ->
-            instrs.(pc') <- Osr {label; cond = c::cond; target; map}
+          | Osr {label; cond; target; var_map; frame_maps} ->
+              instrs.(pc') <- Osr {label; cond = c::cond; target; var_map; frame_maps}
```
Nitpick: alignment

o- replied:
thanks, somehow merlin update changed identation after `->` from 2 to 4 spaces :/

mhyee commented on transform_assumption.ml:
```diff
@@ -32,11 +32,11 @@ let insert_checkpoints (func:afunction) =
-        InsertBefore [Osr {label=checkpoint_label pc; cond=[]; target; map=osr};]
+        InsertBefore [Osr {label=checkpoint_label pc; cond=[]; target; var_map=osr; frame_maps=[]};]
     in
-    if pc = 0 then Unchanged else
```
Why was this removed?

o- replied:
Why was it there in the first place :) I don't know.

Maybe it was because it could go wrong if the first instruction was a label. But since we forbid this now I don't see a reason for this case.

mhyee replied:
OK, I forgot that we don't allow the first instruction to be a label.

mhyee commented on transform_assumption.ml:
```diff
@@ -32,11 +32,11 @@ let insert_checkpoints (func:afunction) =
-    if pc = 0 then Unchanged else
     match[@warning "-4"] instrs.(pc) with
-    | Stop _ | Return _ | Label _ | Comment _ -> Unchanged
+    | Label _ | Comment _ ->
```
Why was this changed?

o- replied:
To have more safepoints available. I noticed that in the case of:
```
call x = foo
return
```
there was no safepoint after the call. but we need one. Not sure if there was ever a good reason for the list.

mhyee commented on transform_inline.ml:
```diff
@@ -154,108 +144,103 @@ let inline ({main; functions} as orig_prog : program) : program option =
+        let callee = apply next in
+        match needs_osr callee.instrs, osr with
+        | false, _ ->
+            let inlinee, new_used = compose cur callee !used_labels ret args in
```
Nitpick: alignment

mhyee commented on transform_inline.ml:
```diff
@@ -154,108 +144,103 @@ let inline ({main; functions} as orig_prog : program) : program option =
+    let inline_order = compute_inline_order func LabelSet.empty in
+    (* If there are no caller-callee pairs to inline, return `None`, else return
+       the completely inlined program *)
+    if inline_order = [] then None
```
```ocaml
if inline_order = []
then None
else
  ...
```

mhyee commented on transform_inline.ml:
```diff
@@ -22,6 +12,9 @@ type inlining_candidate = {
 }
 and inlining_site = inlining_candidate list
 
+let as_inlining_input (func : afunction) =
```
This is basically `Analysis.as_analysis_input` except that it uses the active version, rather than a given version.

mhyee replied:
I would probably just remove this and use `Analysis.as_analysis_input`.

mhyee commented on transform_inline.ml:
```diff
@@ -162,18 +155,31 @@ let inline ({main; functions} as orig_prog : program) : program option =
+        version = version.label;
+        pos = osr_label; } in
+      { var_map = top_frame;
+        continuation_res=call_var;
```
Nitpick: missing space before/after `=`. Also on the line below.

mhyee commented on transform_inline.ml:
```diff
@@ -194,6 +200,19 @@ let inline ({main; functions} as orig_prog : program) : program option =
     Array.exists is_osr
   in
 
+  let fixup_osr extra_frames input =
```
Can you add a short comment documenting this?

mhyee commented on transform_inline.ml:
```diff
@@ -211,9 +230,9 @@ let inline ({main; functions} as orig_prog : program) : program option =
-            (* TODO(osr): Here we need to update the inlinee osr points to create the
-             *            additional osr frame. *)
-            (pos, 0, [| |])
+            let inlinee, new_used = compose cur callee !used_labels ret args in
```
Nitpick: alignment

mhyee commented on transform_inline.ml:
```diff
@@ -154,108 +137,129 @@ let inline ({main; functions} as orig_prog : program) : program option =
+            let inlinee, new_used = compose cur callee !used_labels ret args in
+            used_labels := new_used;
+            (pos, 1, fixup_osr osr inlinee)
+        | true , None ->
```
Nitpick: extra space before `,`

mhyee reviewed (commented):
This looks pretty good! I'll have to look at this again, when I'm less tired and have also looked at the spec in #129 to refresh my memory.
Tests would also be nice.

Also, as a general comment, the semantics doc has fallen behind. We should update it soon. (But it's good that the inlining spec is somewhere, even if it's a PR #129.)

o- committed:
change analysis_input to be a list of variables (not a set)

o- committed:
add doc describing inlining

o- commented:
Ok I reworked it quite a bit after all your suggestions.

Also I made one final change to the osr extra frames. In the spirit of my original proposal the syntax is now:

```
osr sp1 [cond] (target) [top_frame]  (cont1) [var ret1 = $, frame1], (cont2) [var ret2 = $, frame2]
```

where 
* `sp1` is the label
* `cond` is a list of trigger conditions
* `target` is the target tripple where control resumes
* `top_frame` is a variable mapping for the top most frame

the rest is a list of additional frames to create where
* `contx` is a target tripple of the return address
* `retx` is the name of the variable that receives the result of the call (ie. $ is the 'return register')
* `framex` is the variable mapping to create the additional frame

o- committed:
fix order of osr extra frames

o- committed:
fix remove_empty_osr

since now we also jump back to non-local targets we need to finally
extend the remove_empty_osr pass to be a whole program analysis.

o- committed:
add two examples

to see inlining of osr in action run those with optimizations
prune,inline,const_fold

mhyee commented on transform_inline.ml:
```diff
@@ -187,30 +199,63 @@ let inline ({main; functions} as orig_prog : program) : program option =
+   *    (f0,v1,t0) [var res = $, frame0/res], (f1,v1,t1) [frame1], (f2,v2,t2) [frame2]
+   *
+   * This ensures that when the deoptimized foo returns it returns after the call with the call
+   * stack identical to before the call and the result stored to y.
```
`y` -> `res` ?

o- committed:
different inlining heuristics

using 3 different inlining strategies:

inline_small : only small methods and depth 2. enabled in `--opts all`
inline_med   : medium methods, depth 4. enabled in test_examples.sh
inline_max   : inline everything

this allows us to finally run the vector example by default.

mhyee commented on examples/inline_osr.sou:
```diff
@@ -0,0 +1,15 @@
+function main ()
```
It would be nice to include an "after" version so that we don't have to run the program. Unless it's too large?

mhyee replied:
Or at the very least, a comment explaining which opts to run (since we won't be able to dig out the commit message).

o- replied:
not sure if I understand the concerns? This is run by the test runner, that was the main concern. Any optimizations combined with inline (obvious from the file name) is interesting. Why an after? I can just run it...

mhyee replied:
Pick your favourite reason :)

- Many of our interesting examples have a before/after
- Run once and add it to repo, instead of running it each time I want to see it
- Maybe I want to show someone on my phone
- But really, I'm lazy

o- replied:
ok

o- replied:
(of course it revealed a bug)

mhyee commented on transform.ml:
```diff
@@ -119,8 +121,12 @@ let optimize (opts : string list) (prog : program) : program option =
       as_opt_program branch_prune
-    | "inline" ->
-      Transform_inline.inline
+    | "inline_max" ->
```
There are some typos in the commit message (or code):

`inlining_*` -> `inline_*`
`inlining_map` -> `inline_max`

Is `inline_med` to a depth of 3 or 4?

o- replied:
fixed (and there are people saying nobody reads commit msgs :) )

mhyee commented on transform.ml:
```diff
@@ -119,8 +121,12 @@ let optimize (opts : string list) (prog : program) : program option =
+    | "inline_max" ->
+      Transform_inline.inline ()
+    | "inline_med" ->
+      Transform_inline.inline ~max_depth:4 ~max_size:220 ()
```
Just curious, but where did the values 170 and 220 come from?

o- replied:
170 - 220 = 390 which is 10 from 400 which is a round number

o- replied:
(ie. out of thin air)

mhyee commented on transform_assumption.ml:
```diff
@@ -47,17 +47,64 @@ let insert_checkpoints (func:afunction) =
+(* Removes all empty checkpoints with no incoming references *)
+let remove_empty_osr : opt_prog = fun prog ->
+  (* Collect the targets (including the extra frames continuations) of all non empty osrs *)
+  let used =
```
Could you use different names for the different variations of `used`? I would find it easier to follow.

mhyee replied:
At least the definition of `add` is directly above its use. That I can live with.

o- replied:
Yeah, I could already hear your comment when I wrote this ;)

Isn't this the perfect case to get you used to overloading names. Due to the very obvious indentation structure I honestly don't think this would become more clear to read if I invented silly names for all the used steps... I can make you happy and make all of them anonymous :)

mhyee replied:
Anonymous is much worse. :)

Can you rename the first one (line 60) to something like `used_checkpoints` (`used_targets`, etc)? Unlike the others, it's used all the way down on line 88, and it isn't obvious from the indentation.

o- replied:
I have an alternative version that I like. But I am not sure if it is easier to understand now...

o- replied:
The version now is in a sense nicer, but there is also beauty in stupid solutions

mhyee commented on transform_assumption.ml:
```diff
@@ -47,17 +47,64 @@ let insert_checkpoints (func:afunction) =
+
+  (* Remove all empty and unused osr instructions in all versions *)
+  let changed = ref false in
+  let transform prog =
```
Same with `transform`.

mhyee replied:
I'll live with this one.

mhyee reviewed (commented):
I like the new syntax for frames. `(f,v,t,res)` always seemed too similar to `(f,v,t)` for me.

Do you know how long the vector example took under the small and medium inlining? (If you don't have the numbers, I'll just run it myself later.)

o- commented:
> Do you know how long the vector example took under the small and medium inlining? (If you don't have the numbers, I'll just run it myself later.)

about 4-5' for small and a bit more than a minute for the long test runner

Update: ups sorry. those numbers are for the whole testsuite including vector

mhyee commented on doc/inlining.md:
````diff
@@ -0,0 +1,125 @@
+                   (main, base, sp1) [var y = $, var x = x]
+```
+
+The special % register specifies the target binding for the continuation.
````
Typo: `%` -> `$`

o- committed:
fix loading of code with osr instructions already present

o- commented:
ptal

gasche commented on parser.mly:
```diff
@@ -145,8 +151,9 @@ instruction:
-  { Osr {label; cond; target= {func; version; pos}; map} }
+  top=osr_map
+  frames=separated_list(COMMA, osr_frame)
+  { Osr {label; cond; target= {func; version; pos}; varmap=top; frame_maps=frames} }
```
I would rather like the first frame to also be separated from the osr_map by a comma. You can do this easily (and this can simplify `osr_frame` above as well) by just using `list(preceded(COMMA, foo))` instead of `separated_list(COMMA, foo)`. You can also give this pattern a name:

```
%inline preparated_list(sep, elem): list(preceded(sep, elem)) { $1 }
```

o- replied:
thank you, I was looking for that trick!

o- commented:
@gasche are you happy with the rest?

gasche commented:
I haven't reviewed everything in detail, but I think the proposed design is solid and I am fine with merging the PR.
(I noticed a "it's => its" typo somewhere if you care.)

One merely stylistic comment: I think that multi-line `then/else` branches without `begin end` enclosing are dangerous (they work just fine as long as they start with a `let`, but they can surprise you if you start using sequences within their body), I think it would be good style to use `begin end` more often in these situations. No need to change the current patch, though.
